<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API Reference · StochasticGene.jl</title><meta name="title" content="API Reference · StochasticGene.jl"/><meta property="og:title" content="API Reference · StochasticGene.jl"/><meta property="twitter:title" content="API Reference · StochasticGene.jl"/><meta name="description" content="Documentation for StochasticGene.jl."/><meta property="og:description" content="Documentation for StochasticGene.jl."/><meta property="twitter:description" content="Documentation for StochasticGene.jl."/><meta property="og:url" content="https://nih-niddk-mbs.github.io/StochasticGene.jl/stable/api/"/><meta property="twitter:url" content="https://nih-niddk-mbs.github.io/StochasticGene.jl/stable/api/"/><link rel="canonical" href="https://nih-niddk-mbs.github.io/StochasticGene.jl/stable/api/"/><script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXXXX-X"></script><script>  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'UA-XXXXXXXXX-X', {'page_path': location.pathname + location.search + location.hash});
</script><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/custom.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">StochasticGene.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../getting_started/">Getting Started</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../examples/">Overview</a></li><li><input class="collapse-toggle" id="menuitem-4-2" type="checkbox"/><label class="tocitem" for="menuitem-4-2"><span class="docs-label">Basic Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../examples/basic_telegraph/">Basic Telegraph Model</a></li><li><a class="tocitem" href="../examples/multi_state_rna/">Multi-State Model</a></li><li><a class="tocitem" href="../examples/rna_histogram/">RNA Histogram Analysis</a></li><li><a class="tocitem" href="../examples/ms2_analysis/">MS2 Reporter Analysis</a></li><li><a class="tocitem" href="../examples/dual_reporter/">Dual Reporter System</a></li><li><a class="tocitem" href="../examples/trace_analysis/">Trace Analysis</a></li><li><a class="tocitem" href="../examples/dwell_time_analysis/">Basic Dwell Times</a></li><li><a class="tocitem" href="../examples/rna_dwell_time/">RNA Dwell Time Analysis</a></li><li><a class="tocitem" href="../examples/rna_onoff/">RNA On-Off Analysis</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-4-3" type="checkbox"/><label class="tocitem" for="menuitem-4-3"><span class="docs-label">Advanced Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../examples/parallel_processing/">Parallel Processing</a></li><li><a class="tocitem" href="../examples/hierarchical_trace/">Hierarchical Trace Analysis</a></li><li><a class="tocitem" href="../examples/joint_trace/">Joint Trace Analysis</a></li><li><a class="tocitem" href="../examples/coupled_models/">Coupled Model Analysis</a></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li class="is-active"><a class="tocitem" href>API Reference</a><ul class="internal"><li><a class="tocitem" href="#Core-Functions"><span>Core Functions</span></a></li><li><a class="tocitem" href="#Comprehensive-Function-Documentation"><span>Comprehensive Function Documentation</span></a></li><li><a class="tocitem" href="#Data-Types"><span>Data Types</span></a></li><li><a class="tocitem" href="#Model-Types"><span>Model Types</span></a></li><li><a class="tocitem" href="#Utility-Functions"><span>Utility Functions</span></a></li><li><a class="tocitem" href="#Analysis-Functions"><span>Analysis Functions</span></a></li><li><a class="tocitem" href="#Hierarchical-and-Coupling-Models"><span>Hierarchical and Coupling Models</span></a></li><li><a class="tocitem" href="#Model-Components"><span>Model Components</span></a></li><li><a class="tocitem" href="#Supported-Data-Types"><span>Supported Data Types</span></a></li><li><a class="tocitem" href="#Error-Handling"><span>Error Handling</span></a></li><li><a class="tocitem" href="#Performance-Considerations"><span>Performance Considerations</span></a></li></ul></li><li><input class="collapse-toggle" id="menuitem-5-2" type="checkbox"/><label class="tocitem" for="menuitem-5-2"><span class="docs-label">Core Functions</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="fit/">Model Fitting</a></li><li><a class="tocitem" href="simulator/">Model Simulation</a></li><li><a class="tocitem" href="load_data/">Data Loading</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-5-3" type="checkbox"/><label class="tocitem" for="menuitem-5-3"><span class="docs-label">Function Libraries</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="utilities/">Utility Functions</a></li><li><a class="tocitem" href="analysis/">Analysis Functions</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-5-4" type="checkbox"/><label class="tocitem" for="menuitem-5-4"><span class="docs-label">Output Functions</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="write_traces/">Write Traces</a></li><li><a class="tocitem" href="write_ONOFFhistograms/">Write ON/OFF Histograms</a></li><li><a class="tocitem" href="write_residency_G_folder/">Write G State Residency</a></li></ul></li></ul></li><li><a class="tocitem" href="../contributing/">Contributing</a></li></ul><div class="docs-version-selector field has-addons visible"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"><option value="#" selected="selected">1.6.0</option></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>API Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API Reference</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/nih-niddk-mbs/StochasticGene.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/nih-niddk-mbs/StochasticGene.jl/blob/main/docs/src/api/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="API-Reference"><a class="docs-heading-anchor" href="#API-Reference">API Reference</a><a id="API-Reference-1"></a><a class="docs-heading-anchor-permalink" href="#API-Reference" title="Permalink"></a></h1><h2 id="Core-Functions"><a class="docs-heading-anchor" href="#Core-Functions">Core Functions</a><a id="Core-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Functions" title="Permalink"></a></h2><h3 id="Model-Fitting"><a class="docs-heading-anchor" href="#Model-Fitting">Model Fitting</a><a id="Model-Fitting-1"></a><a class="docs-heading-anchor-permalink" href="#Model-Fitting" title="Permalink"></a></h3><ul><li><a href="fit/"><code>fit</code></a>: Main function for fitting models to data</li><li><a href="analysis/#loglikelihood"><code>loglikelihood</code></a>: Calculate log-likelihood for model parameters</li><li><a href="@ref"><code>run_mh</code></a>: Run Metropolis-Hastings MCMC sampling</li><li><a href="@ref"><code>run_mcmc_parallel</code></a>: Run parallel MCMC chains</li></ul><h3 id="Model-Simulation"><a class="docs-heading-anchor" href="#Model-Simulation">Model Simulation</a><a id="Model-Simulation-1"></a><a class="docs-heading-anchor-permalink" href="#Model-Simulation" title="Permalink"></a></h3><ul><li><a href="simulator/"><code>simulator</code></a>: Simulate stochastic gene expression models</li><li><a href="@ref"><code>simulate_trace</code></a>: Generate intensity traces from models</li><li><a href="@ref"><code>simulate_trace_data</code></a>: Generate trace data with metadata</li><li><a href="@ref"><code>simulate_trace_vector</code></a>: Generate vectors of traces</li><li><a href="@ref"><code>simulate_trials</code></a>: Simulate multiple model realizations</li></ul><h3 id="Data-Loading-and-Management"><a class="docs-heading-anchor" href="#Data-Loading-and-Management">Data Loading and Management</a><a id="Data-Loading-and-Management-1"></a><a class="docs-heading-anchor-permalink" href="#Data-Loading-and-Management" title="Permalink"></a></h3><ul><li><a href="load_data/"><code>load_data</code></a>: Load experimental data from files</li><li><a href="@ref"><code>load_model</code></a>: Load model parameters from files</li><li><a href="@ref"><code>rna_setup</code></a>: Set up project directory structure</li><li><a href="@ref"><code>readrates</code></a>: Read rate parameters from files</li><li><a href="@ref"><code>readfile</code></a>: Read data files with error handling</li></ul><h3 id="Analysis-and-Visualization"><a class="docs-heading-anchor" href="#Analysis-and-Visualization">Analysis and Visualization</a><a id="Analysis-and-Visualization-1"></a><a class="docs-heading-anchor-permalink" href="#Analysis-and-Visualization" title="Permalink"></a></h3><ul><li><a href="write_traces/"><code>write_traces</code></a>: Generate model-predicted intensity traces</li><li><a href="write_ONOFFhistograms/"><code>write_ONOFFhistograms</code></a>: Generate ON/OFF dwell time histograms</li><li><a href="write_residency_G_folder/"><code>write_residency_G_folder</code></a>: Generate G state residency probabilities</li><li><a href="@ref"><code>write_histograms</code></a>: Write RNA histogram predictions</li><li><a href="@ref"><code>write_dataframes</code></a>: Write results to CSV files</li></ul><h2 id="Comprehensive-Function-Documentation"><a class="docs-heading-anchor" href="#Comprehensive-Function-Documentation">Comprehensive Function Documentation</a><a id="Comprehensive-Function-Documentation-1"></a><a class="docs-heading-anchor-permalink" href="#Comprehensive-Function-Documentation" title="Permalink"></a></h2><h3 id="Core-Function-Libraries"><a class="docs-heading-anchor" href="#Core-Function-Libraries">Core Function Libraries</a><a id="Core-Function-Libraries-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Function-Libraries" title="Permalink"></a></h3><ul><li><a href="utilities/"><code>Utilities</code></a>: Data processing, model construction, and utility functions</li><li><a href="analysis/"><code>Analysis</code></a>: Post-fitting analysis, model comparison, and visualization functions</li></ul><h2 id="Data-Types"><a class="docs-heading-anchor" href="#Data-Types">Data Types</a><a id="Data-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Data-Types" title="Permalink"></a></h2><h3 id="RNA-Data-Structures"><a class="docs-heading-anchor" href="#RNA-Data-Structures">RNA Data Structures</a><a id="RNA-Data-Structures-1"></a><a class="docs-heading-anchor-permalink" href="#RNA-Data-Structures" title="Permalink"></a></h3><h4 id="RNAData"><a class="docs-heading-anchor" href="#RNAData">RNAData</a><a id="RNAData-1"></a><a class="docs-heading-anchor-permalink" href="#RNAData" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct RNAData{nType,hType} &lt;: AbstractRNAData{hType}
    label::String     # Data set label
    gene::String      # Gene name (case sensitive)
    nRNA::nType       # Length of histogram
    histRNA::hType    # RNA histogram data
end</code></pre><p>A structure for storing steady-state RNA count distributions from techniques like smFISH or scRNA-seq.</p><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Create RNA data from histogram
data = RNAData(
    &quot;control&quot;,          # label
    &quot;MYC&quot;,             # gene
    50,                # nRNA
    [10,20,30,25,15]   # histRNA
)</code></pre><h4 id="RNACountData"><a class="docs-heading-anchor" href="#RNACountData">RNACountData</a><a id="RNACountData-1"></a><a class="docs-heading-anchor-permalink" href="#RNACountData" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct RNACountData &lt;: AbstractRNAData{Vector{Int}}
    label::String
    gene::String
    nRNA::Int
    countsRNA::Vector{Int}
    yieldfactor::Vector{Float64}
end</code></pre><p>A structure for storing individual RNA count measurements with yield factors.</p><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Create RNA count data with yield correction
data = RNACountData(
    &quot;single_cell&quot;,     # label
    &quot;ACTB&quot;,           # gene
    100,              # nRNA
    [1,2,3,4,5],      # countsRNA
    [0.8,0.9,0.85]    # yieldfactor
)</code></pre><h4 id="RNAOnOffData"><a class="docs-heading-anchor" href="#RNAOnOffData">RNAOnOffData</a><a id="RNAOnOffData-1"></a><a class="docs-heading-anchor-permalink" href="#RNAOnOffData" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct RNAOnOffData &lt;: AbstractHistogramData
    label::String
    gene::String
    nRNA::Int
    histRNA::Vector
    bins::Vector
    ON::Vector        # ON time probability density
    OFF::Vector       # OFF time probability density
end</code></pre><p>A structure for storing combined RNA count and ON/OFF state duration data.</p><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Create combined RNA and ON/OFF data
data = RNAOnOffData(
    &quot;live_cell&quot;,      # label
    &quot;SOX2&quot;,          # gene
    30,              # nRNA
    [5,10,15,20],    # histRNA
    [0,1,2,3,4],     # bins
    [0.1,0.3,0.4,0.2], # ON
    [0.2,0.4,0.3,0.1]  # OFF
)</code></pre><h4 id="RNADwellTimeData"><a class="docs-heading-anchor" href="#RNADwellTimeData">RNADwellTimeData</a><a id="RNADwellTimeData-1"></a><a class="docs-heading-anchor-permalink" href="#RNADwellTimeData" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct RNADwellTimeData &lt;: AbstractHistogramData
    label::String
    gene::String
    nRNA::Int
    histRNA::Array
    bins::Vector{Vector}
    DwellTimes::Vector{Vector}
    DTtypes::Vector
end</code></pre><p>A structure for storing RNA counts with dwell time distributions.</p><h3 id="Trace-Data-Structures"><a class="docs-heading-anchor" href="#Trace-Data-Structures">Trace Data Structures</a><a id="Trace-Data-Structures-1"></a><a class="docs-heading-anchor-permalink" href="#Trace-Data-Structures" title="Permalink"></a></h3><h4 id="TraceData"><a class="docs-heading-anchor" href="#TraceData">TraceData</a><a id="TraceData-1"></a><a class="docs-heading-anchor-permalink" href="#TraceData" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct TraceData{labelType,geneType,traceType} &lt;: AbstractTraceData
    label::labelType    # Data set label
    gene::geneType      # Gene name
    interval::Float64   # Time interval between trace points
    trace::traceType    # Trace data
end</code></pre><p>A structure for storing fluorescence intensity time series data.</p><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Create trace data
traces = [[1.0, 2.0, 3.0], [2.0, 3.0, 4.0]]
data = TraceData(
    &quot;live_imaging&quot;,   # label
    &quot;MYC&quot;,           # gene
    1.0,             # interval (minutes)
    traces           # trace data
)</code></pre><h4 id="TraceRNAData"><a class="docs-heading-anchor" href="#TraceRNAData">TraceRNAData</a><a id="TraceRNAData-1"></a><a class="docs-heading-anchor-permalink" href="#TraceRNAData" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct TraceRNAData{traceType,hType} &lt;: AbstractTraceHistogramData
    label::String
    gene::String
    interval::Float64
    trace::traceType
    nRNA::Int
    histRNA::hType
end</code></pre><p>A structure for storing both trace and RNA histogram data.</p><h4 id="DwellTimeData"><a class="docs-heading-anchor" href="#DwellTimeData">DwellTimeData</a><a id="DwellTimeData-1"></a><a class="docs-heading-anchor-permalink" href="#DwellTimeData" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct DwellTimeData &lt;: AbstractHistogramData
    label::String
    gene::String
    bins::Vector
    DwellTimes::Vector
    DTtypes::Vector
end</code></pre><p>A structure for storing dwell time distributions only.</p><h2 id="Model-Types"><a class="docs-heading-anchor" href="#Model-Types">Model Types</a><a id="Model-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Model-Types" title="Permalink"></a></h2><h3 id="Gene-Model-(GM)"><a class="docs-heading-anchor" href="#Gene-Model-(GM)">Gene Model (GM)</a><a id="Gene-Model-(GM)-1"></a><a class="docs-heading-anchor-permalink" href="#Gene-Model-(GM)" title="Permalink"></a></h3><h4 id="GMmodel"><a class="docs-heading-anchor" href="#GMmodel">GMmodel</a><a id="GMmodel-1"></a><a class="docs-heading-anchor-permalink" href="#GMmodel" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct GMmodel{RateType,PriorType,ProposalType,ParamType,MethodType,ComponentType,ReporterType} &lt;: AbstractGMmodel
    rates::RateType              # Transition rates
    Gtransitions::Tuple          # G state transitions
    G::Int                       # Number of G states
    nalleles::Int               # Number of alleles
    rateprior::PriorType        # Rate prior distributions
    proposal::ProposalType      # MCMC proposal distribution
    fittedparam::ParamType      # Fitted parameter indices
    fixedeffects::Tuple         # Fixed effects specification
    method::MethodType          # Solution method
    components::ComponentType   # Model components
    reporter::ReporterType      # Reporter configuration
end</code></pre><p>A structure for Gene (G) models with arbitrary numbers of gene states.</p><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Create a simple two-state gene model
model = GMmodel(
    rates = [0.1, 0.2],                    # G1-&gt;G2, G2-&gt;G1
    Gtransitions = ([1,2], [2,1]),         # State transitions
    G = 2,                                 # Two gene states
    nalleles = 2,                          # Diploid
    # ... other parameters
)</code></pre><h3 id="Gene-Reporter-Splice-Model-(GRSM)"><a class="docs-heading-anchor" href="#Gene-Reporter-Splice-Model-(GRSM)">Gene-Reporter-Splice Model (GRSM)</a><a id="Gene-Reporter-Splice-Model-(GRSM)-1"></a><a class="docs-heading-anchor-permalink" href="#Gene-Reporter-Splice-Model-(GRSM)" title="Permalink"></a></h3><h4 id="GRSMmodel"><a class="docs-heading-anchor" href="#GRSMmodel">GRSMmodel</a><a id="GRSMmodel-1"></a><a class="docs-heading-anchor-permalink" href="#GRSMmodel" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct GRSMmodel{TraitType,RateType,nratesType,GType,PriorType,ProposalType,ParamType,MethodType,ComponentType,ReporterType} &lt;: AbstractGRSMmodel{TraitType}
    trait::TraitType            # Model traits (hierarchical, coupling, etc.)
    rates::RateType             # Transition rates
    transforms::Transformation  # Rate transformations
    nrates::nratesType          # Number of rates
    Gtransitions::Tuple         # G state transitions
    G::GType                    # Number of G states
    R::GType                    # Number of R (pre-RNA) steps
    S::GType                    # Number of S (splice) sites
    insertstep::GType           # Reporter insertion step
    nalleles::Int               # Number of alleles
    splicetype::String          # Splicing type
    rateprior::PriorType        # Rate prior distributions
    proposal::ProposalType      # MCMC proposal distribution
    fittedparam::ParamType      # Fitted parameter indices
    fixedeffects::Tuple         # Fixed effects specification
    method::MethodType          # Solution method
    components::ComponentType   # Model components
    reporter::ReporterType      # Reporter configuration
end</code></pre><p>A comprehensive structure for Gene-Reporter-Splice models with arbitrary numbers of gene states (G), pre-RNA steps (R), and splice sites (S).</p><p><strong>Example:</strong></p><pre><code class="language-julia hljs"># Create a GRS model with 2 gene states, 3 pre-RNA steps, 2 splice sites
model = GRSMmodel(
    rates = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6],  # G, R, S transitions
    Gtransitions = ([1,2], [2,1]),            # G state transitions
    G = 2,                                    # Two gene states
    R = 3,                                    # Three pre-RNA steps
    S = 2,                                    # Two splice sites
    insertstep = 1,                           # Reporter visible from step 1
    nalleles = 2,                             # Diploid
    splicetype = &quot;offeject&quot;,                  # Splicing type
    # ... other parameters
)</code></pre><h3 id="Supporting-Types"><a class="docs-heading-anchor" href="#Supporting-Types">Supporting Types</a><a id="Supporting-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Supporting-Types" title="Permalink"></a></h3><h4 id="HMMReporter"><a class="docs-heading-anchor" href="#HMMReporter">HMMReporter</a><a id="HMMReporter-1"></a><a class="docs-heading-anchor-permalink" href="#HMMReporter" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct HMMReporter
    n::Int                      # Number of noise parameters
    per_state::Vector           # Reporters per state
    probfn::Function           # Noise distribution function
    weightind::Int             # Mixture weight index
    offstates::Vector{Int}     # Off states
    noiseparams::Vector{Int}   # Noise parameter indices
end</code></pre><p>A structure for configuring reporter properties in Hidden Markov Models.</p><h4 id="Transformation"><a class="docs-heading-anchor" href="#Transformation">Transformation</a><a id="Transformation-1"></a><a class="docs-heading-anchor-permalink" href="#Transformation" title="Permalink"></a></h4><pre><code class="language-julia hljs">struct Transformation
    f::Vector{Function}        # Forward transformations
    f_inv::Vector{Function}    # Inverse transformations
    f_cv::Vector{Function}     # CV transformations
end</code></pre><p>A structure for parameter transformations during fitting.</p><h2 id="Utility-Functions"><a class="docs-heading-anchor" href="#Utility-Functions">Utility Functions</a><a id="Utility-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Utility-Functions" title="Permalink"></a></h2><h3 id="Data-Processing"><a class="docs-heading-anchor" href="#Data-Processing">Data Processing</a><a id="Data-Processing-1"></a><a class="docs-heading-anchor-permalink" href="#Data-Processing" title="Permalink"></a></h3><ul><li><a href="utilities/#normalize_histogram"><code>normalize_histogram</code></a>: Normalize probability distributions</li><li><a href="utilities/#make_array"><code>make_array</code></a>: Convert data to arrays</li><li><a href="utilities/#make_mat"><code>make_mat</code></a>: Convert data to matrices</li><li><a href="utilities/#digit_vector"><code>digit_vector</code></a>: Convert numbers to digit vectors</li></ul><h3 id="Model-Construction"><a class="docs-heading-anchor" href="#Model-Construction">Model Construction</a><a id="Model-Construction-1"></a><a class="docs-heading-anchor-permalink" href="#Model-Construction" title="Permalink"></a></h3><ul><li><a href="utilities/#prepare_rates"><code>prepare_rates</code></a>: Prepare rate parameters for fitting</li><li><a href="utilities/#get_rates"><code>get_rates</code></a>: Extract rates from fitted parameters</li><li><a href="utilities/#get_param"><code>get_param</code></a>: Extract specific parameters</li><li><a href="utilities/#num_rates"><code>num_rates</code></a>: Count number of rates in model</li><li><a href="@ref"><code>num_all_parameters</code></a>: Count total parameters</li></ul><h3 id="Statistical-Functions"><a class="docs-heading-anchor" href="#Statistical-Functions">Statistical Functions</a><a id="Statistical-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Statistical-Functions" title="Permalink"></a></h3><ul><li><a href="utilities/#prob_Gaussian"><code>prob_Gaussian</code></a>: Gaussian probability density</li><li><a href="@ref"><code>prob_Gaussian_grid</code></a>: Gaussian probability on grid</li><li><a href="utilities/#mean_elongationtime"><code>mean_elongationtime</code></a>: Calculate mean elongation time</li><li><a href="utilities/#on_states"><code>on_states</code></a>: Identify transcriptionally active states</li><li><a href="utilities/#source_states"><code>source_states</code></a>: Identify source states for transitions</li></ul><h3 id="File-I/O"><a class="docs-heading-anchor" href="#File-I/O">File I/O</a><a id="File-I/O-1"></a><a class="docs-heading-anchor-permalink" href="#File-I/O" title="Permalink"></a></h3><ul><li><a href="utilities/#folder_path"><code>folder_path</code></a>: Construct folder paths</li><li><a href="utilities/#folder_setup"><code>folder_setup</code></a>: Set up directory structure</li><li><a href="utilities/#datapdf"><code>datapdf</code></a>: Generate data PDFs</li><li><a href="utilities/#make_dataframes"><code>make_dataframes</code></a>: Create DataFrames from results</li><li><a href="@ref"><code>write_dataframes_only</code></a>: Write DataFrames to files</li></ul><h2 id="Analysis-Functions"><a class="docs-heading-anchor" href="#Analysis-Functions">Analysis Functions</a><a id="Analysis-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Analysis-Functions" title="Permalink"></a></h2><h3 id="Model-Diagnostics"><a class="docs-heading-anchor" href="#Model-Diagnostics">Model Diagnostics</a><a id="Model-Diagnostics-1"></a><a class="docs-heading-anchor-permalink" href="#Model-Diagnostics" title="Permalink"></a></h3><ul><li><a href="utilities/#large_deviance"><code>large_deviance</code></a>: Identify chains with large deviance</li><li><a href="utilities/#large_rhat"><code>large_rhat</code></a>: Identify parameters with large R-hat</li><li><a href="@ref"><code>assemble_measures_model</code></a>: Assemble model measures</li><li><a href="utilities/#assemble_all"><code>assemble_all</code></a>: Assemble all results</li></ul><h3 id="Post-fitting-Analysis"><a class="docs-heading-anchor" href="#Post-fitting-Analysis">Post-fitting Analysis</a><a id="Post-fitting-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Post-fitting-Analysis" title="Permalink"></a></h3><ul><li><a href="analysis/#predictedarray"><code>predictedarray</code></a>: Generate model predictions</li><li><a href="analysis/#predictedfn"><code>predictedfn</code></a>: Generate prediction functions</li><li><a href="analysis/#make_traces"><code>make_traces</code></a>: Generate trace predictions</li><li><a href="@ref"><code>make_traces_dataframe</code></a>: Convert traces to DataFrames</li></ul><h3 id="Visualization-Support"><a class="docs-heading-anchor" href="#Visualization-Support">Visualization Support</a><a id="Visualization-Support-1"></a><a class="docs-heading-anchor-permalink" href="#Visualization-Support" title="Permalink"></a></h3><ul><li><a href="@ref"><code>write_cov</code></a>: Write covariance matrices</li><li><a href="@ref"><code>write_augmented</code></a>: Write augmented results</li><li><a href="@ref"><code>write_winners</code></a>: Write best-fit results</li></ul><h2 id="Hierarchical-and-Coupling-Models"><a class="docs-heading-anchor" href="#Hierarchical-and-Coupling-Models">Hierarchical and Coupling Models</a><a id="Hierarchical-and-Coupling-Models-1"></a><a class="docs-heading-anchor-permalink" href="#Hierarchical-and-Coupling-Models" title="Permalink"></a></h2><h3 id="Hierarchical-Models"><a class="docs-heading-anchor" href="#Hierarchical-Models">Hierarchical Models</a><a id="Hierarchical-Models-1"></a><a class="docs-heading-anchor-permalink" href="#Hierarchical-Models" title="Permalink"></a></h3><p>For hierarchical models, use the <code>hierarchical</code> parameter in <code>fit()</code>:</p><pre><code class="language-julia hljs"># Fit hierarchical model
fits = fit(
    hierarchical = (2, [1,2]),  # 2 hyperparameter sets, fit rates 1,2
    # ... other parameters
)</code></pre><h3 id="Coupled-Models"><a class="docs-heading-anchor" href="#Coupled-Models">Coupled Models</a><a id="Coupled-Models-1"></a><a class="docs-heading-anchor-permalink" href="#Coupled-Models" title="Permalink"></a></h3><p>For coupled transcriptional units, use tuples for model parameters:</p><pre><code class="language-julia hljs"># Fit coupled model with different G states
fits = fit(
    G = (2, 3),                          # Unit 1: 2 states, Unit 2: 3 states
    transitions = (([1,2], [2,1]), ([1,2], [2,3], [3,1])),
    coupling = (1, 2, [1,2], [3,4], 2),  # Coupling specification
    # ... other parameters
)</code></pre><h2 id="Model-Components"><a class="docs-heading-anchor" href="#Model-Components">Model Components</a><a id="Model-Components-1"></a><a class="docs-heading-anchor-permalink" href="#Model-Components" title="Permalink"></a></h2><h3 id="Gene-States-(G)"><a class="docs-heading-anchor" href="#Gene-States-(G)">Gene States (G)</a><a id="Gene-States-(G)-1"></a><a class="docs-heading-anchor-permalink" href="#Gene-States-(G)" title="Permalink"></a></h3><ul><li>Arbitrary number of gene states</li><li>User-specified transitions between states</li><li>One active state for transcription initiation</li><li>Support for multiple alleles</li></ul><h3 id="Pre-RNA-Steps-(R)"><a class="docs-heading-anchor" href="#Pre-RNA-Steps-(R)">Pre-RNA Steps (R)</a><a id="Pre-RNA-Steps-(R)-1"></a><a class="docs-heading-anchor-permalink" href="#Pre-RNA-Steps-(R)" title="Permalink"></a></h3><ul><li>Irreversible forward transitions</li><li>mRNA ejection from final R step</li><li>Optional reporter insertion at any step</li><li>Support for elongation dynamics</li></ul><h3 id="Splicing-(S)"><a class="docs-heading-anchor" href="#Splicing-(S)">Splicing (S)</a><a id="Splicing-(S)-1"></a><a class="docs-heading-anchor-permalink" href="#Splicing-(S)" title="Permalink"></a></h3><ul><li>Up to R splice sites</li><li>PreRNA with or without spliced intron</li><li>Multiple configurations per R step</li><li>Support for different splicing types:<ul><li><code>&quot;&quot;</code>: No splicing</li><li><code>&quot;offeject&quot;</code>: Splice then eject</li><li><code>&quot;offdecay&quot;</code>: Splice then decay</li></ul></li></ul><h2 id="Supported-Data-Types"><a class="docs-heading-anchor" href="#Supported-Data-Types">Supported Data Types</a><a id="Supported-Data-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Supported-Data-Types" title="Permalink"></a></h2><p>The package handles multiple experimental data types:</p><ol><li><p><strong>mRNA Count Distributions</strong></p><ul><li>Single molecule FISH (smFISH)</li><li>Single cell RNA sequencing (scRNA-seq)</li><li>Bulk RNA measurements</li></ul></li><li><p><strong>Intensity Traces</strong></p><ul><li>Live cell fluorescence microscopy</li><li>Time-lapse imaging</li><li>Multiple reporter constructs</li></ul></li><li><p><strong>Dwell Time Distributions</strong></p><ul><li>ON/OFF state durations</li><li>Transcriptional burst analysis</li><li>Reporter visibility periods</li></ul></li><li><p><strong>Combined Data Types</strong></p><ul><li>RNA + trace data</li><li>RNA + dwell time data</li><li>Multi-modal experiments</li></ul></li></ol><h2 id="Error-Handling"><a class="docs-heading-anchor" href="#Error-Handling">Error Handling</a><a id="Error-Handling-1"></a><a class="docs-heading-anchor-permalink" href="#Error-Handling" title="Permalink"></a></h2><p>The package includes comprehensive error handling:</p><ul><li><strong>Data validation</strong>: Checks for consistent data formats</li><li><strong>Parameter validation</strong>: Ensures valid model specifications</li><li><strong>Convergence monitoring</strong>: Tracks MCMC convergence</li><li><strong>Memory management</strong>: Handles large datasets efficiently</li></ul><h2 id="Performance-Considerations"><a class="docs-heading-anchor" href="#Performance-Considerations">Performance Considerations</a><a id="Performance-Considerations-1"></a><a class="docs-heading-anchor-permalink" href="#Performance-Considerations" title="Permalink"></a></h2><ul><li><strong>Parallel processing</strong>: Use <code>nchains &gt; 1</code> for parallel MCMC</li><li><strong>Memory usage</strong>: Large datasets may require cluster computing</li><li><strong>Convergence</strong>: Monitor R-hat values for chain convergence</li><li><strong>Optimization</strong>: Use appropriate priors for faster convergence</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../examples/coupled_models/">« Coupled Model Analysis</a><a class="docs-footer-nextpage" href="fit/">Model Fitting »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.11.4 on <span class="colophon-date" title="Friday 6 February 2026 16:21">Friday 6 February 2026</span>. Using Julia version 1.11.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
