<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Getting Started · StochasticGene.jl</title><meta name="title" content="Getting Started · StochasticGene.jl"/><meta property="og:title" content="Getting Started · StochasticGene.jl"/><meta property="twitter:title" content="Getting Started · StochasticGene.jl"/><meta name="description" content="Documentation for StochasticGene.jl."/><meta property="og:description" content="Documentation for StochasticGene.jl."/><meta property="twitter:description" content="Documentation for StochasticGene.jl."/><meta property="og:url" content="https://nih-niddk-mbs.github.io/StochasticGene.jl/stable/getting_started/"/><meta property="twitter:url" content="https://nih-niddk-mbs.github.io/StochasticGene.jl/stable/getting_started/"/><link rel="canonical" href="https://nih-niddk-mbs.github.io/StochasticGene.jl/stable/getting_started/"/><script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXXXX-X"></script><script>  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'UA-XXXXXXXXX-X', {'page_path': location.pathname + location.search + location.hash});
</script><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/custom.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">StochasticGene.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../installation/">Installation</a></li><li class="is-active"><a class="tocitem" href>Getting Started</a><ul class="internal"><li><a class="tocitem" href="#Folder-Structure"><span>Folder Structure</span></a></li><li><a class="tocitem" href="#Setting-Up"><span>Setting Up</span></a></li><li><a class="tocitem" href="#Data-Preparation"><span>Data Preparation</span></a></li><li><a class="tocitem" href="#Fitting-a-Model"><span>Fitting a Model</span></a></li><li><a class="tocitem" href="#Model-Components"><span>Model Components</span></a></li><li><a class="tocitem" href="#Model-Types"><span>Model Types</span></a></li><li><a class="tocitem" href="#Data-Types"><span>Data Types</span></a></li><li><a class="tocitem" href="#Next-Steps"><span>Next Steps</span></a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../examples/">Overview</a></li><li><input class="collapse-toggle" id="menuitem-4-2" type="checkbox"/><label class="tocitem" for="menuitem-4-2"><span class="docs-label">Basic Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../examples/basic_telegraph/">Basic Telegraph Model</a></li><li><a class="tocitem" href="../examples/multi_state_rna/">Multi-State Model</a></li><li><a class="tocitem" href="../examples/rna_histogram/">RNA Histogram Analysis</a></li><li><a class="tocitem" href="../examples/ms2_analysis/">MS2 Reporter Analysis</a></li><li><a class="tocitem" href="../examples/dual_reporter/">Dual Reporter System</a></li><li><a class="tocitem" href="../examples/trace_analysis/">Trace Analysis</a></li><li><a class="tocitem" href="../examples/dwell_time_analysis/">Basic Dwell Times</a></li><li><a class="tocitem" href="../examples/rna_dwell_time/">RNA Dwell Time Analysis</a></li><li><a class="tocitem" href="../examples/rna_onoff/">RNA On-Off Analysis</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-4-3" type="checkbox"/><label class="tocitem" for="menuitem-4-3"><span class="docs-label">Advanced Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../examples/parallel_processing/">Parallel Processing</a></li><li><a class="tocitem" href="../examples/hierarchical_trace/">Hierarchical Trace Analysis</a></li><li><a class="tocitem" href="../examples/joint_trace/">Joint Trace Analysis</a></li><li><a class="tocitem" href="../examples/coupled_models/">Coupled Model Analysis</a></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../api/">API Reference</a></li><li><input class="collapse-toggle" id="menuitem-5-2" type="checkbox"/><label class="tocitem" for="menuitem-5-2"><span class="docs-label">Core Functions</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../api/fit/">Model Fitting</a></li><li><a class="tocitem" href="../api/simulator/">Model Simulation</a></li><li><a class="tocitem" href="../api/load_data/">Data Loading</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-5-3" type="checkbox"/><label class="tocitem" for="menuitem-5-3"><span class="docs-label">Function Libraries</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../api/utilities/">Utility Functions</a></li><li><a class="tocitem" href="../api/analysis/">Analysis Functions</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-5-4" type="checkbox"/><label class="tocitem" for="menuitem-5-4"><span class="docs-label">Output Functions</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../api/write_traces/">Write Traces</a></li><li><a class="tocitem" href="../api/write_ONOFFhistograms/">Write ON/OFF Histograms</a></li><li><a class="tocitem" href="../api/write_residency_G_folder/">Write G State Residency</a></li></ul></li></ul></li><li><a class="tocitem" href="../contributing/">Contributing</a></li></ul><div class="docs-version-selector field has-addons visible"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"><option value="#" selected="selected">1.6.0</option></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Getting Started</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Getting Started</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/nih-niddk-mbs/StochasticGene.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/nih-niddk-mbs/StochasticGene.jl/blob/main/docs/src/getting_started.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Getting-Started"><a class="docs-heading-anchor" href="#Getting-Started">Getting Started</a><a id="Getting-Started-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-Started" title="Permalink"></a></h1><h2 id="Folder-Structure"><a class="docs-heading-anchor" href="#Folder-Structure">Folder Structure</a><a id="Folder-Structure-1"></a><a class="docs-heading-anchor-permalink" href="#Folder-Structure" title="Permalink"></a></h2><p>StochasticGene expects a specific directory structure for data and results:</p><pre><code class="language-bash hljs">project_root/
├── data/
│   ├── alleles/      # Contains allele count information
│   └── halflives/    # Contains mRNA half-life information
└── results/          # Output directory for analysis results</code></pre><h2 id="Setting-Up"><a class="docs-heading-anchor" href="#Setting-Up">Setting Up</a><a id="Setting-Up-1"></a><a class="docs-heading-anchor-permalink" href="#Setting-Up" title="Permalink"></a></h2><p>Use the built-in helper to set up a new project folder with the required data structure and reference files:</p><pre><code class="language-julia hljs">using StochasticGene
rna_setup()           # Sets up folders in the current directory
# or, to set up in a subfolder:
rna_setup(&quot;scRNA&quot;)    # Sets up folders in ./scRNA
cd(&quot;scRNA&quot;)           # Move into the project directory (if you want to work there)</code></pre><p>This will create the necessary folders and download reference CSV files for alleles and halflives. It does not generate arbitrary mock data, but provides the standard structure and example reference files needed to get started.</p><h2 id="Data-Preparation"><a class="docs-heading-anchor" href="#Data-Preparation">Data Preparation</a><a id="Data-Preparation-1"></a><a class="docs-heading-anchor-permalink" href="#Data-Preparation" title="Permalink"></a></h2><h3 id="RNA-Count-Data"><a class="docs-heading-anchor" href="#RNA-Count-Data">RNA Count Data</a><a id="RNA-Count-Data-1"></a><a class="docs-heading-anchor-permalink" href="#RNA-Count-Data" title="Permalink"></a></h3><p>RNA count data should be provided as a text file with the following format:</p><pre><code class="language-text hljs"># gene_condition.txt
count1
count2
count3
...</code></pre><p>The data will be loaded into an <code>RNAData</code> structure with the following fields:</p><ul><li><code>histRNA</code>: Vector of RNA counts</li><li><code>gene</code>: Gene name</li><li><code>condition</code>: Experimental condition</li><li><code>nalleles</code>: Number of alleles</li></ul><h3 id="Live-Cell-Imaging-Data"><a class="docs-heading-anchor" href="#Live-Cell-Imaging-Data">Live Cell Imaging Data</a><a id="Live-Cell-Imaging-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Live-Cell-Imaging-Data" title="Permalink"></a></h3><p>For live cell imaging data, provide a text file with time series data:</p><pre><code class="language-text hljs"># gene_condition.txt
time1 intensity1
time2 intensity2
time3 intensity3
...</code></pre><h2 id="Fitting-a-Model"><a class="docs-heading-anchor" href="#Fitting-a-Model">Fitting a Model</a><a id="Fitting-a-Model-1"></a><a class="docs-heading-anchor-permalink" href="#Fitting-a-Model" title="Permalink"></a></h2><h3 id="Basic-Two-State-Model"><a class="docs-heading-anchor" href="#Basic-Two-State-Model">Basic Two-State Model</a><a id="Basic-Two-State-Model-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-Two-State-Model" title="Permalink"></a></h3><p>Fit a simple two-state model to RNA count data:</p><pre><code class="language-julia hljs">fits = fit(
    G = 2,                    # Two gene states
    R = 0,                    # No pre-RNA steps
    transitions = ([1,2], [2,1]),  # Transitions between states
    datatype = &quot;rna&quot;,         # RNA count data
    datapath = &quot;data/HCT116_testdata/&quot;,
    gene = &quot;MYC&quot;,            # Gene name
    datacond = &quot;MOCK&quot;        # Experimental condition
)</code></pre><h3 id="Model-with-Pre-RNA-Steps"><a class="docs-heading-anchor" href="#Model-with-Pre-RNA-Steps">Model with Pre-RNA Steps</a><a id="Model-with-Pre-RNA-Steps-1"></a><a class="docs-heading-anchor-permalink" href="#Model-with-Pre-RNA-Steps" title="Permalink"></a></h3><p>Fit a model with pre-RNA processing steps:</p><pre><code class="language-julia hljs">fits = fit(
    G = 2,                    # Two gene states
    R = 3,                    # Three pre-RNA steps
    S = 2,                    # Two splice sites
    insertstep = 1,           # Reporter insertion at step 1
    transitions = ([1,2], [2,1]),  # Gene state transitions
    datatype = &quot;trace&quot;,       # Live cell imaging data
    datapath = &quot;data/testtraces&quot;,
    gene = &quot;MS2&quot;,            # Gene name
    datacond = &quot;testtrace&quot;,   # Experimental condition
    traceinfo = (1.0, 1., -1, 1.),  # Frame interval, start, end, active fraction
    noisepriors = [40., 20., 200., 10.],  # Noise parameters
    nchains = 4              # Number of MCMC chains
)</code></pre><h2 id="Model-Components"><a class="docs-heading-anchor" href="#Model-Components">Model Components</a><a id="Model-Components-1"></a><a class="docs-heading-anchor-permalink" href="#Model-Components" title="Permalink"></a></h2><h3 id="Gene-States-(G)"><a class="docs-heading-anchor" href="#Gene-States-(G)">Gene States (G)</a><a id="Gene-States-(G)-1"></a><a class="docs-heading-anchor-permalink" href="#Gene-States-(G)" title="Permalink"></a></h3><ul><li>Arbitrary number of gene states</li><li>User-specified transitions between states</li><li>One active state for transcription initiation</li><li>Example transitions:<ul><li>Two-state: <code>([1,2], [2,1])</code> (telegraph model)</li><li>Three-state: <code>([1,2], [2,1], [2,3], [3,1])</code> (cyclic model)</li></ul></li></ul><h3 id="Pre-RNA-Steps-(R)"><a class="docs-heading-anchor" href="#Pre-RNA-Steps-(R)">Pre-RNA Steps (R)</a><a id="Pre-RNA-Steps-(R)-1"></a><a class="docs-heading-anchor-permalink" href="#Pre-RNA-Steps-(R)" title="Permalink"></a></h3><ul><li>Irreversible forward transitions</li><li>mRNA ejection from final R step</li><li>Optional reporter insertion step</li><li>Example: R=3 with reporter at step 1:<pre><code class="language-julia hljs">R = 3
insertstep = 1</code></pre></li></ul><h3 id="Splicing-(S)"><a class="docs-heading-anchor" href="#Splicing-(S)">Splicing (S)</a><a id="Splicing-(S)-1"></a><a class="docs-heading-anchor-permalink" href="#Splicing-(S)" title="Permalink"></a></h3><ul><li>Up to R splice sites</li><li>PreRNA with or without spliced intron</li><li>Multiple configurations per R step</li><li>Example: S=2 with R=3:<pre><code class="language-julia hljs">R = 3
S = 2</code></pre></li></ul><h2 id="Model-Types"><a class="docs-heading-anchor" href="#Model-Types">Model Types</a><a id="Model-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Model-Types" title="Permalink"></a></h2><p>StochasticGene supports several model types:</p><h3 id="G-Models-(Telegraph-Models)"><a class="docs-heading-anchor" href="#G-Models-(Telegraph-Models)">G Models (Telegraph Models)</a><a id="G-Models-(Telegraph-Models)-1"></a><a class="docs-heading-anchor-permalink" href="#G-Models-(Telegraph-Models)" title="Permalink"></a></h3><ul><li>Basic two-state model</li><li>Multiple gene states</li><li>Example:<pre><code class="language-julia hljs">G = 2
R = 0
transitions = ([1,2], [2,1])</code></pre></li></ul><h3 id="GR-Models-(with-Pre-RNA-Steps)"><a class="docs-heading-anchor" href="#GR-Models-(with-Pre-RNA-Steps)">GR Models (with Pre-RNA Steps)</a><a id="GR-Models-(with-Pre-RNA-Steps)-1"></a><a class="docs-heading-anchor-permalink" href="#GR-Models-(with-Pre-RNA-Steps)" title="Permalink"></a></h3><ul><li>Gene states plus pre-RNA processing</li><li>Example:<pre><code class="language-julia hljs">G = 2
R = 3
transitions = ([1,2], [2,1])</code></pre></li></ul><h3 id="GRS-Models-(with-Splicing)"><a class="docs-heading-anchor" href="#GRS-Models-(with-Splicing)">GRS Models (with Splicing)</a><a id="GRS-Models-(with-Splicing)-1"></a><a class="docs-heading-anchor-permalink" href="#GRS-Models-(with-Splicing)" title="Permalink"></a></h3><ul><li>Gene states, pre-RNA steps, and splicing</li><li>Example:<pre><code class="language-julia hljs">G = 2
R = 3
S = 2
insertstep = 1</code></pre></li></ul><h3 id="Coupled-Models-(Multiple-Alleles)"><a class="docs-heading-anchor" href="#Coupled-Models-(Multiple-Alleles)">Coupled Models (Multiple Alleles)</a><a id="Coupled-Models-(Multiple-Alleles)-1"></a><a class="docs-heading-anchor-permalink" href="#Coupled-Models-(Multiple-Alleles)" title="Permalink"></a></h3><ul><li>Multiple alleles with coupling</li><li>Example:<pre><code class="language-julia hljs">G = (2,2)  # Two alleles, each with 2 states
R = (3,3)  # Three pre-RNA steps for each
coupling = (
    (1,1),  # Model indices
    ([2], [1]),  # Source units
    ([&quot;G2&quot;], [&quot;G1&quot;]),  # Source states
    ([2], [1])  # Target transitions
)</code></pre></li></ul><h2 id="Data-Types"><a class="docs-heading-anchor" href="#Data-Types">Data Types</a><a id="Data-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Data-Types" title="Permalink"></a></h2><p>The package can handle:</p><h3 id="RNA-Count-Data-2"><a class="docs-heading-anchor" href="#RNA-Count-Data-2">RNA Count Data</a><a class="docs-heading-anchor-permalink" href="#RNA-Count-Data-2" title="Permalink"></a></h3><ul><li>Single molecule FISH (smFISH)</li><li>Single cell RNA sequencing (scRNA-seq)</li><li>Example:<pre><code class="language-julia hljs">datatype = &quot;rna&quot;
datapath = &quot;data/rna_counts/&quot;</code></pre></li></ul><h3 id="Live-Cell-Imaging"><a class="docs-heading-anchor" href="#Live-Cell-Imaging">Live Cell Imaging</a><a id="Live-Cell-Imaging-1"></a><a class="docs-heading-anchor-permalink" href="#Live-Cell-Imaging" title="Permalink"></a></h3><ul><li>MS2 reporter data</li><li>PP7 reporter data</li><li>Example:<pre><code class="language-julia hljs">datatype = &quot;trace&quot;
datapath = &quot;data/ms2_traces/&quot;
traceinfo = (1.0, 1., -1, 1.)  # Frame interval, start, end, active fraction</code></pre></li></ul><h3 id="Dwell-Time-Analysis"><a class="docs-heading-anchor" href="#Dwell-Time-Analysis">Dwell Time Analysis</a><a id="Dwell-Time-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Dwell-Time-Analysis" title="Permalink"></a></h3><ul><li>ON/OFF state durations</li><li>Example:<pre><code class="language-julia hljs">datatype = &quot;dwelltime&quot;
datapath = &quot;data/dwell_times/&quot;</code></pre></li></ul><h3 id="Combined-Data"><a class="docs-heading-anchor" href="#Combined-Data">Combined Data</a><a id="Combined-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Combined-Data" title="Permalink"></a></h3><ul><li>RNA counts with dwell times</li><li>Example:<pre><code class="language-julia hljs">datatype = &quot;rnadwelltime&quot;
datapath = [&quot;data/rna_counts/&quot;, &quot;data/dwell_times/&quot;]</code></pre></li></ul><h2 id="Next-Steps"><a class="docs-heading-anchor" href="#Next-Steps">Next Steps</a><a id="Next-Steps-1"></a><a class="docs-heading-anchor-permalink" href="#Next-Steps" title="Permalink"></a></h2><ul><li>Check the <a href="@ref">Examples</a> section for more complex usage scenarios</li><li>Read the <a href="../api/#API-Reference">API Reference</a> for detailed function documentation</li><li>Join the <a href="https://github.com/nih-niddk-mbs/StochasticGene.jl/discussions">GitHub discussions</a> for community support</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../installation/">« Installation</a><a class="docs-footer-nextpage" href="../examples/">Overview »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.11.4 on <span class="colophon-date" title="Friday 6 February 2026 16:21">Friday 6 February 2026</span>. Using Julia version 1.11.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
